angular/projects/admin/src/app/services/document.service.ts

```diff
@@ -111,15 +118,30 @@ export class DocumentService {
         query = query.orderBy(queryOpts.orderBy.field, queryOpts.orderBy.sortOrder);
       }
 
+      if (!nextPage && backDocSnap) { // back
+        query = query.startAt(backDocSnap);
+      }
+
       if (queryOpts.limit) {
         query = query.limit(queryOpts.limit);
       }
 
+      if (nextPage && lastDocSnap) { // next page
+        query = query.startAfter(lastDocSnap);
+      }
       return query;
     };
 
     return this.siteCollection
-      .collection<Document>('documents', queryFn)
-      .valueChanges();
+    .collection<Document>('documents', queryFn).valueChanges();
+  }
+
+  getReference(id: string) {
```

Comment:
Might not be needed with the changes to `query()` above.

Reply:
getReference still needed to get documentSnaphot for querying startAfter

Reply:
How about the list of results to `DocumentListDataSource`? Can do we have them as an array there where we can get the first and last snapshot?

Reply:
DocumentListDataSource only return an Observable of document data. All Snapshot metadata is stripped. This method provides only the data. So we need using snapShotChanges to get metadata that contain document reference

https://github.com/angular/angularfire2/blob/master/docs/firestore/documents.md

> valueChanges()
What is it? - Returns an Observable of document data. All Snapshot metadata is stripped. This method provides only the data.
>
> Why would you use it? - When you just need the object data. No document metadata is attached which makes it simple to render to a view.
>
> When would you not use it? - When you need the id of the document to use data manipulation methods. This method assumes you either are saving the id to the document data or using a "readonly" approach.

 
> snapshotChanges()
> What is it? - Returns an Observable of data as a DocumentChangeAction.
> 
> Why would you use it? - When you need the document data but also want to keep around metadata. This metadata provides you the underyling DocumentReference and document id. Having the document's id around makes it easier to use data manipulation methods. This method gives you more horsepower with other Angular integrations such as ngrx, forms, and animations due to the type property. The type property on each DocumentChangeAction is useful for ngrx reducers, form states, and animation states.

Reply:
Arright, I see. Thanks for that.